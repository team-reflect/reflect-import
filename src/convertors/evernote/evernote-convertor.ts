import {parseHtml, removeImgsWithDataSrcs} from 'helpers/html'
import {validateNotes} from 'helpers/validate'
import {parseXml} from 'helpers/xml'

import {parseTime, toEvernoteId} from './evernote-helpers'
import {ConvertedNote, ConvertOptions, Convertor, ConvertResponse} from '../../types'

export class EvernoteConvertor implements Convertor {
  accept = {'application/enex': ['.enex']}

  convert({data}: ConvertOptions): ConvertResponse {
    const doc = parseXml(data)

    const noteDocs = Array.from(doc.querySelectorAll('en-export > note'))

    const notes = noteDocs.map((noteDoc, index) => this.convertNoteDoc(noteDoc, index))

    return validateNotes(notes)
  }

  private convertNoteDoc(noteDoc: Element, index: number): ConvertedNote {
    const subject = this.extractSubject(noteDoc)
    const html = this.extractHtml(noteDoc)
    const timestamps = this.extractTimestamps(noteDoc)
    const id = toEvernoteId(index, subject)

    return {id, html, subject, ...timestamps}
  }

  private extractSubject(noteDoc: Element): string | undefined {
    const subject = noteDoc.querySelector('title')?.textContent ?? undefined
    return subject
  }

  private extractHtml(noteDoc: Element): string {
    const content = noteDoc.querySelector('content')?.textContent ?? ''
    const contentDoc = parseXml(content)
    const contentNoteDoc = contentDoc.querySelector('en-note')
    const html = contentNoteDoc?.innerHTML ?? ''
    const doc = parseHtml(html)

    removeImgsWithDataSrcs(doc)

    return doc.body.innerHTML
  }

  private extractTimestamps(noteDoc: Element) {
    const createdAtString = noteDoc.querySelector('created')?.textContent
    const updatedAtString = noteDoc.querySelector('updated')?.textContent

    const createdAt = createdAtString ? parseTime(createdAtString) : undefined
    const updatedAt = updatedAtString ? parseTime(updatedAtString) : undefined

    return {createdAt, updatedAt}
  }
}
