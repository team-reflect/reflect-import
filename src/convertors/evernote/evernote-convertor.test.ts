import {describe, expect, it} from 'vitest'

import {EvernoteConvertor} from './evernote-convertor'

describe('EvernoteConvertor', () => {
  it('converts evernote to HTML', () => {
    const convertor = new EvernoteConvertor()

    const data = `<?xml version="1.0" encoding="UTF-8"?>
  <en-export export-date="20221224T010732Z" application="Evernote" version="10.49.4">
    <note>
      <title>My first note</title>
      <created>20221124T001227Z</created>
      <updated>20221124T010654Z</updated>
      <note-attributes>
      </note-attributes>
      <content>
        <![CDATA[<?xml version="1.0" encoding="UTF-8" standalone="no"?>
  <!DOCTYPE en-note SYSTEM "http://xml.evernote.com/pub/enml2.dtd"><en-note><h2>h2 title</h2><ul><li><div>list <a href="https://google.com">link</a> </div></li></ul></en-note>      ]]>
      </content>
    </note>
  </en-export>
  `
    const {notes} = convertor.convert({data})
    const [{subject, html, createdAt, updatedAt}] = notes

    expect(subject).toEqual('My first note')
    expect(html).toMatchInlineSnapshot(
      '"<h2>h2 title</h2><ul><li><div>list <a href=\\"https://google.com\\">link</a> </div></li></ul>"',
    )

    expect(createdAt).toEqual(new Date('2022-11-24T00:12:27.000Z').getTime())
    expect(updatedAt).toEqual(new Date('2022-11-24T01:06:54.000Z').getTime())
  })

  it('raises an error if parsing fails', () => {
    const convertor = new EvernoteConvertor()

    expect(() => convertor.convert({data: ''})).toThrow()
  })

  it.each(['&gt;', '&nbsp;', '&mdash;', '&ndash;'])(
    'parses note with html entity',
    (entity) => {
      const convertor = new EvernoteConvertor()

      const data = `<?xml version="1.0" encoding="UTF-8"?>
    <!DOCTYPE en-export SYSTEM "http://xml.evernote.com/pub/evernote-export4.dtd">
    <en-export export-date="20221230T101804Z" application="Evernote" version="10.49.4">
      <note>
        <title>failing nbsp</title>
        <created>20191201T063018Z</created>
        <updated>20221130T101752Z</updated>
        <note-attributes>
        </note-attributes>
        <content>
          <![CDATA[<?xml version="1.0" encoding="UTF-8" standalone="no"?><!DOCTYPE en-note SYSTEM "http://xml.evernote.com/pub/enml2.dtd"><en-note><div>foo ${entity} bar</div></en-note>      ]]>
        </content>
      </note>
    </en-export>`

      const {notes} = convertor.convert({data})
      const [{subject, html}] = notes

      expect(subject).toEqual('failing nbsp')
      expect(html).toMatchSnapshot()
    },
  )
})
