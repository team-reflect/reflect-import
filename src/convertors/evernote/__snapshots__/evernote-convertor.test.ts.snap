// Vitest Snapshot v1

exports[`EvernoteConvertor > parses note with html entity 1`] = `"<div>foo &gt; bar</div>"`;

exports[`EvernoteConvertor > parses note with html entity 2`] = `"<div>foo &nbsp; bar</div>"`;
